### DISCUSSION TITLE
No warping on segmentation images

### DISCUSSION THREAD

**Discussion by Kanyq** [voxelmorph/voxelmorph#281] [Q&A]
Hello,

I've trained the model for 3D liver MRI dataset for 100 epochs and it is giving me good moved predicted images but when I try to warp the segmentation images using predicted warping (pred_warp) I **don't get any warping in the segmentation images**.

I'm not sure what's causing this. If someone has faced this or know what's going on and how to fix it then I would really appreciate your input.

Thanks

**Comment by adalca** [comment]
Can you show the command you use to wash an image and to warp the segmentation? Also did you visualize the warp field?

**Reply by Kanyq** [reply]
This is the command I used for warping the segmentation image:
warped_seg_mov = warp_model.predict([segmentation_array[17][np.newaxis,...,np.newaxis], pred_warp])
segmentation_array[17] is the "moving" segmentation image of dimension 288, 288, 96.

I also tried to visualize the warp field but since its a 3D field I can't properly visualize it. As "ne.plot.flow" wants a "2D+2" dimension image. I'm not able to find a proper workaround for it. So any leads on this will also be very helpful.

Even though the flow I generated doesn't seem appropriate I'm attaching it so maybe I can get some advice from you or from anyone on what might be going on here. The dimension for the flow is (96, 32, 2). The original "pred_warp" shape was (1, 288, 288, 96, 3).
![flow](https://user-images.githubusercontent.com/14558146/111497257-3f539b80-8741-11eb-879b-c42b9bfb8d69.png)

**Reply by adalca** [reply]
thanks. it would be really useful if we knew what `warp_model` is (how it was construted) and if we could see some screenshots of e.g. input/output images and input/output segmentations. Did you also warp the moving image using the same `warp_model`?

**Reply by Kanyq** [reply]
Sure, 
So after training the model we try to predict the "moved image" from a pair of fixed and moving images. So we put these two images(numpy array) into a list.
val_input =[
    fix[np.newaxis,...,np.newaxis],
    mov[np.newaxis,...,np.newaxis]
]

This "val_input" is used to generate the predictions.
val_pred = vxm_model.predict(val_input)

The prediction(val_pred) gives us a list of 2 numpy arrays. 
moved_pred = val_pred[0] 
pred_warp = val_pred[1]

moved_pred - This numpy array gives us the predicted moved image (based on moving and fix image pair we provided in input)
pred_warp - This numpy array gives us the predicted warping

The "warp_model" is the internal Transformer of the voxelmoprh as they have provided in their tutorial. Here's the code of how we define the "warp_model".
warp_model = vxm.networks.Transform(vol_shape, interp_method='nearest')
comment  - vol_shape is the shape of images (288, 288, 96)

Now as you asked I'm providing some images for your reference.

This image shows the fixed image vs. the predicted moved image
![fix_pred](https://user-images.githubusercontent.com/14558146/111507509-08828300-874b-11eb-819a-ad0baa033c32.png)

This image shows the moving image vs. the predicted moved image
![mov_pred](https://user-images.githubusercontent.com/14558146/111507744-44b5e380-874b-11eb-8690-57254c81195e.png)

This image shows the moving segmentation image vs. the warped segmentation image.
![seg](https://user-images.githubusercontent.com/14558146/111508529-197fc400-874c-11eb-94b2-fa665bda851d.png)

This image shows the fixed segmentation image vs. the warped segmentation image. And as we can see the there has been no warping in the "warped segmentation image" to match it to the fixed segmentation image.
![fix_seg](https://user-images.githubusercontent.com/14558146/111508668-3fa56400-874c-11eb-95f0-cd0c86633315.png)

I can also provide the numpy(.npy) files for the predicted 
moved_pred = val_pred[0] 
pred_warp = val_pred[1]

**Reply by adalca** [reply]
I'm a bit worried that you mixed your inputs. vxm warps the first image you pass it, not the second. So it should take [moving, fixed]. So it's likely what's happening here is that the model isn't trained properly?

By the way, depending on the voxelmorph model you use, the second output might be the SVF instead of the deformation field. To deform an image and be sure you do it right, you might use:
`vxm_model.apply_transform(src, tgt, img_to_be_warped, interp_method='nearest')` This warps src to tgt, takes taht deformation field and applies it to `img_to_be_warped`

**Reply by Kanyq** [reply]
Okay, I think what you're saying makes sense and I'll try to train the model for higher epochs and then try to predict the results. Also for transformation I'll use "apply_transform". I will let you know once I've trained the model. And just as a last query for now. When training the model we're providing the data in [fix, mov] format to the model. Or do I need to change it to [mov, fix]?

**Reply by adalca** [reply]
for the default model, you provide two images [A, B], and what it will do internally is warp A and compare it to B.

in your definition, yes i think you need to switch them.

**Reply by Kanyq** [reply]
okay thanks, I'll make the changes as you suggested

**Reply by Kanyq** [reply]
I've currently put the model on training for 100 epochs at 5 steps per epoch and will let you know the results of the training. And thank you so much for your help.

**Reply by Kanyq** [reply]
Hello again,

I trained the model for 20 epochs to see what deformations and predictions I will get. But I'm not seeing any difference between moving and predicted image. As per your suggestion, I changed the image order to [moving, fix] in both training the model and then predicting it.

This image is fixed vs. predicted image
![fixed_moved](https://user-images.githubusercontent.com/14558146/111520108-138fe000-8758-11eb-9430-19f0fcd8d8ec.png)

This image is moving vs. predicted image
![moving_moved](https://user-images.githubusercontent.com/14558146/111520304-433ee800-8758-11eb-9b48-3e39380305c4.png)

I'm not sure what I'm doing wrong can you please guide me on this. Thanks

**Reply by adalca** [reply]
Well 20 epochs is quite limited for large images like this, but most importantly these images don't seem normalized in [0, 1], whereas our normalization parameters expect images to be normalized to [0, 1] or maybe even [0, 0.5]. Could you try that? Also, just as a test, can you make your images smaller (maybe downsample to 96x96x96) just to see if it works. once we know it works, you can go back to full size.

**Reply by Kanyq** [reply]
okay I'll try normalizing it and downsample the images
